File: rust/core/tests/doc_string_diff_test.rs
//! Port of `DocStringDiffTest.java` / `doc-string-diff.test.ts`.

use varar_core::cell_diff::CellDiff;
use varar_core::doc_string_diff::{DOC_STRING_COLUMN, compare_doc_string};
use varar_core::error::StepError;
use varar_core::span::Span;
use varar_core::value::Value;

const SPAN: Span = Span {
    start_offset: 0,
    end_offset: 6,
    start_line: 1,
    start_col: 1,
    end_line: 1,
    end_col: 6,
};

#[test]
fn compare_doc_string_equal_content_returns_null() {
    assert_eq!(None, compare_doc_string(Some(&Value::from("hello\n")), "hello\n", SPAN).unwrap());
}

#[test]
fn compare_doc_string_null_return_returns_null_asserted_nothing() {
    assert_eq!(None, compare_doc_string(None, "hello\n", SPAN).unwrap());
}

#[test]
fn compare_doc_string_different_content_returns_a_cell_labelled_doc_string() {
    // A doc string is one cell, compared whole. expected/actual are quoted so a
    // whitespace-only difference stays visible.
    let diff = compare_doc_string(Some(&Value::from("bye\n")), "hello\n", SPAN)
        .unwrap()
        .unwrap();
    assert_eq!(DOC_STRING_COLUMN, diff.column);
    assert_eq!(SPAN, diff.span);
    assert_eq!("\"hello\\n\"", diff.expected);
    assert_eq!("\"bye\\n\"", diff.actual);
    assert!(!diff.ok);
}

#[test]
fn a_doc_string_cell_reads_like_any_other_cell_mismatch() {
    let diff: CellDiff = compare_doc_string(Some(&Value::from("bye\n")), "hello\n", SPAN)
        .unwrap()
        .unwrap();
    assert_eq!(
        "doc string: expected \"hello\\n\" but was \"bye\\n\"",
        StepError::CellMismatch(vec![diff]).message()
    );
}

#[test]
fn compare_doc_string_a_non_string_return_throws_return_shape() {
    assert!(matches!(
        compare_doc_string(Some(&Value::Int(42)), "hello\n", SPAN).unwrap_err(),
        StepError::ReturnShape(_)
    ));
}

File: java/core/src/test/java/dev/varar/core/DocStringDiffTest.java
package dev.varar.core;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.junit.jupiter.api.Test;

/** Translated from {@code var-core/tests/doc-string-diff.test.ts}. */
class DocStringDiffTest {

    private static final Span SPAN = new Span(0, 6, 1, 1, 1, 6);

    @Test
    void compareDocStringEqualContentReturnsNull() {
        assertNull(DocStringDiff.compareDocString("hello\n", "hello\n", SPAN));
    }

    @Test
    void compareDocStringNullReturnReturnsNullAssertedNothing() {
        assertNull(DocStringDiff.compareDocString(null, "hello\n", SPAN));
    }

    @Test
    void compareDocStringDifferentContentReturnsACellLabelledDocString() {
        // A doc string is one cell, compared whole. expected/actual are quoted so a
        // whitespace-only difference stays visible.
        CellDiff diff = DocStringDiff.compareDocString("bye\n", "hello\n", SPAN);
        assertEquals(DocStringDiff.DOC_STRING_COLUMN, diff.column());
        assertEquals(SPAN, diff.span());
        assertEquals("\"hello\\n\"", diff.expected());
        assertEquals("\"bye\\n\"", diff.actual());
        assertFalse(diff.ok());
    }

    @Test
    void aDocStringCellReadsLikeAnyOtherCellMismatch() {
        CellDiff diff = DocStringDiff.compareDocString("bye\n", "hello\n", SPAN);
        assertEquals(
                "doc string: expected \"hello\\n\" but was \"bye\\n\"",
                new CellDiff.CellMismatchException(java.util.List.of(diff)).getMessage());
    }

    @Test
    void compareDocStringANonStringReturnThrowsReturnShapeException() {
        assertThrows(CellDiff.ReturnShapeException.class, () -> DocStringDiff.compareDocString(42, "hello\n", SPAN));
    }
}
